hal/halx86/include/apic.h

```diff
@@ -1,253 +1,301 @@
-#define APIC_LVT_REMOTE_IRR		  (0x1 << 14)	/* Remote IRR */
-#define APIC_LVT_LEVEL_TRIGGER		  (0x1 << 15)	/* Lvel Triggered */
-#define APIC_LVT_MASKED			  (0x1 << 16)   /* Mask */
-#define APIC_LVT_PERIODIC		  (0x1 << 17)   /* Timer Mode */
-
-#define APIC_LVT3_DM        (0x7 << 8)
-#define APIC_LVT3_IIPP      (0x1 << 13)
-#define APIC_LVT3_TM        (0x1 << 15)
-#define APIC_LVT3_MASKED    (0x1 << 16)
-#define APIC_LVT3_OS        (0x1 << 17)
-
-#define APIC_TDCR_TMBASE   (0x1 << 2)
-#define APIC_TDCR_MASK     0x0F
-#define APIC_TDCR_2        0x00
-#define APIC_TDCR_4        0x01
-#define APIC_TDCR_8        0x02
-#define APIC_TDCR_16       0x03
-#define APIC_TDCR_32       0x08
-#define APIC_TDCR_64       0x09
-#define APIC_TDCR_128      0x0A
-#define APIC_TDCR_1        0x0B
-
-#define APIC_TARGET_SELF         0x100
-#define APIC_TARGET_ALL          0x200
-#define APIC_TARGET_ALL_BUT_SELF 0x300
-
-#define APIC_INTEGRATED(version) (version & 0xF0)
-
-typedef enum {
-  amPIC = 0,    /* IMCR and PIC compatibility mode */
-  amVWIRE       /* Virtual Wire compatibility mode */
-} APIC_MODE;
-
-#ifdef CONFIG_SMP
-#define MAX_CPU   32
-#else
-#define MAX_CPU	  1
-#endif
+/* The IMCR is supported by two read/writable or write-only I/O ports,
+   22h and 23h, which receive address and data respectively.
+   To access the IMCR, write a value of 70h to I/O port 22h, which selects the IMCR.
+   Then write the data to I/O port 23h. The power-on default value is zero,
+   which connects the NMI and 8259 INTR lines directly to the BSP.
+   Writing a value of 01h forces the NMI and 8259 INTR signals to pass through the APIC.
+*/
+#define IMCR_ADDRESS_PORT  (PUCHAR)0x0022
+#define IMCR_DATA_PORT     (PUCHAR)0x0023
+#define IMCR_SELECT        0x70
+#define IMCR_PIC_DIRECT    0x00
+#define IMCR_PIC_VIA_APIC  0x01
+
+#define ZERO_VECTOR 0x00 //IRQL 00
+#define APC_VECTOR           0x3D // IRQL 01
+#define DISPATCH_VECTOR      0x41 // IRQL 02
+#define APIC_GENERIC_VECTOR  0xC1 // IRQL 27
+#define APIC_CLOCK_VECTOR    0xD1 // IRQL 28
+#define APIC_SYNCH_VECTOR    0xD1 // IRQL 28
+#define APIC_IPI_VECTOR      0xE1 // IRQL 29
+#define APIC_ERROR_VECTOR    0xE3
+#define POWERFAIL_VECTOR     0xEF // IRQL 30
+#define APIC_PROFILE_VECTOR  0xFD // IRQL 31
+#define APIC_PERF_VECTOR     0xFE
+#define APIC_NMI_VECTOR      0xFF
 
-/*
- * Local APIC timer IRQ vector is on a different priority level,
- * to work around the 'lost local interrupt if more than 2 IRQ
- * sources per level' errata.
- */
-#define LOCAL_TIMER_VECTOR 	    0xEF
-
-#define IPI_VECTOR		    0xFB
-#define ERROR_VECTOR		    0xFE
-#define SPURIOUS_VECTOR		    0xFF  /* Must be 0xXF */
-
-/* CPU flags */
-#define CPU_USABLE   0x01  /* 1 if the CPU is usable (ie. can be used) */
-#define CPU_ENABLED  0x02  /* 1 if the CPU is enabled */
-#define CPU_BSP      0x04  /* 1 if the CPU is the bootstrap processor */
-#define CPU_TSC      0x08  /* 1 if the CPU has a time stamp counter */
-
-typedef struct _CPU_INFO
```

Comment:
You think this structure will not be useful?

Reply:
This won't end up being used.